--- gcc/ada/gnatbind.adb.orig	2018-01-11 09:55:25.000000000 +0100
+++ gcc/ada/gnatbind.adb	2019-03-10 14:12:19.136219885 +0100
@@ -439,6 +439,11 @@
          then
             Add_Lib_Search_Dir (Argv (4 .. Argv'Last));
 
+         --  -freestanding
+
+         elsif Argv (2 .. Argv'Last) = "freestanding" then
+            Opt.Freestanding := True;
+
          --  -nostdlib
 
          elsif Argv (2 .. Argv'Last) = "nostdlib" then
--- gcc/ada/bindgen.adb.orig	2018-01-11 09:55:25.000000000 +0100
+++ gcc/ada/bindgen.adb	2019-03-10 14:12:19.140219889 +0100
@@ -1696,6 +1696,9 @@
             WBI ("   function Ada_Main_Program return Integer;");
          else
             WBI ("   procedure Ada_Main_Program;");
+            if Opt.Freestanding then
+               WBI ("   pragma No_Return (Ada_Main_Program);");
+            end if;
          end if;
 
          Set_String ("   pragma Import (Ada, Ada_Main_Program, """);
@@ -1807,7 +1810,7 @@
       --  with a pragma Volatile in order to tell the compiler to preserve
       --  this variable at any level of optimization.
 
-      if Bind_Main_Program and not CodePeer_Mode then
+      if Bind_Main_Program and not CodePeer_Mode and not Opt.Freestanding then
          WBI ("      Ensure_Reference : aliased System.Address := " &
               "Ada_Main_Program_Name'Address;");
          WBI ("      pragma Volatile (Ensure_Reference);");
@@ -2426,6 +2429,10 @@
             end if;
          end if;
 
+         if Opt.Freestanding then
+            WBI ("   pragma No_Return (" & Get_Main_Name & ");");
+         end if;
+
          WBI ("   pragma Export (C, " & Get_Main_Name & ", """ &
            Get_Main_Name & """);");
       end if;
--- gcc/ada/opt.ads.orig	2018-01-11 09:55:25.000000000 +0100
+++ gcc/ada/opt.ads	2019-03-10 14:12:19.140219889 +0100
@@ -729,6 +729,13 @@
    --  the flag -gnatN (which is now considered obsolescent, since the GCC
    --  back end can do a better job of inlining than the front end these days.
 
+   Freestanding : Boolean := False;
+   --  GNATBIND
+   --  Set to True if no environment exists. This is typically the
+   --  case in embedded systems without an operating system. The main
+   --  program does not read command line parameters nor does it
+   --  return.
+
    Full_Path_Name_For_Brief_Errors : Boolean := False;
    --  PROJECT MANAGER
    --  When True, in Brief_Output mode, each error message line will start with
